IRSA-1524: FinderChart fail to launch using API

loitly opened the pull request:
Ticket: https://caltech-ipac.atlassian.net/browse/IRSA-1524
Changes are also in `ife` under the same branch name.

Also Fixes:
- json parameters were not encoded
- can't limit by survey/source
- rename source to sources to match original documented launch api
- encoding/decoding issues in QueryUtils which causes existing api to fail
- expose searchCatalog parameter
- Refactor FinderChartApi code

PR deployed to k8s: https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/

Below are sample tested URLs.  I suggest trying some on your own as well.

Based on original documentation:
https://irsa.ipac.caltech.edu/onlinehelp/finderchart/api.html

// all defaults using UserTargetWorldPt
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?mode=getResult&UserTargetWorldPt=202.48417;47.23056;EQ_J2000

// all defaults using RA, DEC
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?mode=getResult&RA=202.48417&DEC=47.23056

// all defaults using locstr
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?mode=getResult&locstr=m51
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?mode=getResult&locstr=00:42:44.3+-41:16:08

// with size and sources
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?mode=getResult&UserTargetWorldPt=202.48417;47.23056;EQ_J2000&subsize=5&sources=WISE

// with size and survey
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?mode=getResult&UserTargetWorldPt=202.48417;47.23056;EQ_J2000&subsize=5&survey=WISE,2MASS

// without catalog
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?mode=getResult&UserTargetWorldPt=202.48417;47.23056;EQ_J2000&subsize=5&survey=WISE,2MASS&searchCatalog=no

// bad requests...
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?mode=xxx
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?mode=getResult&locstr=xxx
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?mode=getResult&UserTargetWorldPt=202.48417;47.23056;EQ_J2000&sources=xxx

// some sample of the existing api
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=m101
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=m101&survey=2mass&survey=wise 
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=m101&survey=2mass&reproject=true 
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=m101&marker_color=yellow&marker=true&marker_size=20&page_layout=OnePerSurvey 
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?mode=getImage&file_type=pdf&subsetsize=5&survey=DSS%2CSDSS%2C2MASS%2CIRIS%2CWISE&locstr=291.468658447%2B40.343852997

ejoliet commented:
I've tested it from a local build i did because i got weird problem and issues when doing the test via the kubernetes build.
All the links works so i would declare victory and agree to merge after Wei or Lijun test it independently.

Once merge, we can ask Luisa to test it again in test.

Thank you for fixing the issues. Looks everything has been taking care.

loitly commented:
> I've tested it from a local build i did because i got weird problem and issues when doing the test via the kubernetes build.

Sorry, I fail to mention that DSS searches were not returning at the time I submitted this PR.  I assumed they would have fixed it by the time you're reviewing it.  DSS image search is back online now and the version on k8s is working okay for me.

lznakano commented:
Tested many cases based on the doc and the list above, all work well.

lrebull commented:
If I use http rather than https, I get errors:
 rebull% wget --output-document=out.xml "http://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=m101"
--2018-03-06 09:05:56--  http://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=m101
Resolving irsawebdev9.ipac.caltech.edu... 134.4.70.136
Connecting to irsawebdev9.ipac.caltech.edu|134.4.70.136|:80... connected.
HTTP request sent, awaiting response... 308 Permanent Redirect
2018-03-06 09:05:56 ERROR 308: Permanent Redirect.

lrebull commented:
(i'm proceeding using --no-check-certificate)

loitly commented:
> --no-check-certificate

@lrebull  Yes, this flag is needed for now.  We will either get a real certificate or switch back to HTTP in the future.

lrebull commented:
OK, so yesterday @ejoliet told me, "all the examples you had before should work now." Nope. One has moved from "used to work the last time I tested, a few days ago" to "fails"; four have moved from "fails" to "now works". Here are my raw notes:

works:
wget --no-check-certificate --output-document=out.xml "https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=m83"
wget --no-check-certificate --output-document=out.xml "https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=NGC+7479"
wget --no-check-certificate --output-document=out.xml "https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=23h04m56.63s%2B12d19m22.7s+Equ+J2000"
wget --no-check-certificate --output-document=out.xml "https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=22h57m57.5s+%2B26d09m00.09s+Equatorial+B1950"
wget --no-check-certificate --output-document=out.xml "https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=17h44m34s+-27d59m13s"
wget --no-check-certificate --output-document=out.xml "https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=00h42m44.3s+%2B41d16m08s+b1950"
wget --no-check-certificate --output-document=out.xml "https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=00:42:44.3+-41:16:08"
wget --no-check-certificate --output-document=out.xml "https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=34.5565,54.2321 gal"
wget --no-check-certificate --output-document=out.xml "https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=34.5565,54.2321+gal"
wget --no-check-certificate --output-document=out.xml "https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=34.+54.+ecl"
wget --no-check-certificate --output-document=out.xml "https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=34.,+54. ecl"
NEWLY working:
wget --no-check-certificate --output-document=out.xml "https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=34.5565%2B54.2321+gal"
wget --no-check-certificate --output-document=out.xml "https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=34.%2B54.+ecl"
wget --no-check-certificate --output-document=out.xml "https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=34.5565+54.2321+gal"
wget --no-check-certificate --output-document=out.xml "https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=34.%2B54.+ecl"


doesn't:
NEWLY not working:
 wget --no-check-certificate --output-document=out.xml "https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=2MASXJ23045666%2B1219223"
still broken:
wget --no-check-certificate --output-document=out.xml "https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=17h+44m+34s+-27d+59m+13s+Equatorial+J2000"
wget --no-check-certificate --output-document=out.xml "https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=00+42+44.3+-41+16+08+b1983"
wget --no-check-certificate --output-document=out.xml "https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=004244.3+-411608"
wget --no-check-certificate --output-document=out.xml "https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=17h+-27d+Equ+J2000"
wget --no-check-certificate --output-document=out.xml "https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=17h+-27d"

lrebull commented:
This is working as expected (except for WISE's catalog, which is down for other reasons):
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=m83&file_type=png&mode=getResult

Similarly as to above:
works:
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?mode=getResult&locstr=m83
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?mode=getResult&locstr=NGC+7479
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?mode=getResult&locstr=23h04m56.63s%2B12d19m22.7s+Equ+J2000
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?mode=getResult&locstr=22h57m57.5s+%2B26d09m00.09s+Equatorial+B1950
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?mode=getResult&locstr=17h44m34s+-27d59m13s
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?mode=getResult&locstr=00h42m44.3s+%2B41d16m08s+b1950
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?mode=getResult&locstr=00:42:44.3+-41:16:08
NEWLY works https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?mode=getResult&locstr=34.%2B54.+ecl
NEWLY works https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?mode=getResult&locstr=34.%2B54.+ecl

(still) doesn't:
NEWLY fails: https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?mode=getResult&locstr=2MASXJ23045666%2B1219223
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?mode=getResult&locstr=17h+44m+34s+-27d+59m+13s+Equatorial+J2000
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?mode=getResult&locstr=00+42+44.3+-41+16+08+b1983
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?mode=getResult&locstr=004244.3+-411608
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?mode=getResult&locstr=17h+-27d+Equ+J2000
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?mode=getResult&locstr=17h+-27d

lrebull commented:
subsetsize works
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=m83&subsetsize=1&mode=getResult
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=m83&subsetsize=10&mode=getResult
and catalogs NOW load when using this keyword.

lrebull commented:
All of these work as expected:
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=m83&survey=DSS&mode=getResult
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=m83&survey=SDSS&mode=getResult
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=m83&survey=2MASS&mode=getResult
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=m83&survey=WISE&mode=getResult
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=m83&survey=SEIP&mode=getResult
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=m83&survey=AKARI&mode=getResult
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=m83&survey=IRAS&mode=getResult
https://irsawebdev9.ipac.caltech.edu/irsa-1524/applications/finderchart/servlet/api?locstr=m83&survey=IRIS&mode=getResult

ejoliet commented:
@loitly @lrebull i will take a look to the url failing. Thanks for testing!

loitly commented:
@ejoliet , I am also looking at it.  I've reverted a line of code I added because it was decoding twice.
I need to go back and evaluate the fail URLs.  At least one has a bad position.  Other may be format we do not currently support.

ejoliet commented:
yes, the one with '+' space shouldn't be a correct position, but somehow 
it was in the old documentation but we won;t support it because in SHA, 
WISE and others app, the name resolver doesn't accept spaces anyway.
The only one truly failing (but work in name resolver in FC/IV) is the 
first one MAS2J...

On 3/6/18 10:48 AM, Loi Ly wrote:
>
> @ejoliet <https://github.com/ejoliet> , I am also looking at it. I've 
> reverted a line of code I added because it was decoding twice.
> I need to go back and evaluate the fail URLs. At least one has a bad 
> position. Other may be format we do not currently support.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/Caltech-IPAC/firefly/pull/557#issuecomment-370886447>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AAvATN3zSWZnRqnCP6iKZUE_pTazrruZks5tbtn1gaJpZM4Sarq5>.
>

-- 
Emmanuel Joliet
NASA/IPAC  Infrared Science Archive
MS 100-22, Caltech, Pasadena, CA 91125
Office: 168
Email:  ejoliet@ipac.caltech.edu
Phone:  (626) 395-1489

loitly commented:
Another commit was pushed to fixed onlinehelp pointing to 'irsa' instead of 'finderchart'.

Below is a summary of the `locstr` pattern cases.
@ejoliet ... Can we update api doc to reflect what's currently supported instead of addressing them now?  We should create a ticket for the last case, though.

supported locstr patterns:
```
m83
NGC+7479
2MASXJ23045666%2B1219223
23h04m56.63s%2B12d19m22.7s+Equ+J2000
22h57m57.5s+%2B26d09m00.09s+Equatorial+B1950
17h44m34s+-27d59m13s
00h42m44.3s+%2B41d16m08s+b1950
00:42:44.3+-41:16:08
```

unsupported patterns:
```
17h+44m+34s+-27d+59m+13s+Equatorial+J2000
00+42+44.3+-41+16+08+b1983
004244.3+-411608
17h+-27d+Equ+J2000
17h+-27d
```
Should support, but is not:
```
34.5565%2B54.2321+gal
34.%2B54.+ecl
```

lrebull commented:
OK, I can get to help on http://irsawebdev5.ipac.caltech.edu/onlinehelp/finderchart now, thanks.

lrebull commented:
Docs updated and checked back into git and built/deployed here:
http://irsawebdev5.ipac.caltech.edu/onlinehelp/finderchart/#id=api.inputparameters
with these  examples of acceptable name/coordinate formats:
<UL>
<LI>m83
<LI>NGC+7479
<LI>2MASXJ23045666%2B1219223
<LI>23h04m56.63s%2B12d19m22.7s+Equ+J2000
<LI>22h57m57.5s+%2B26d09m00.09s+Equatorial+B1950
<LI>17h44m34s+-27d59m13s
<LI>00h42m44.3s+%2B41d16m08s+b1950
<LI>00:42:44.3+-41:16:08
<LI>34.5565,54.2321 gal
<LI>34.5565,54.2321+gal
<LI>34.5565%2B54.2321+gal
<LI>34.+54.+ecl
<LI>34.+54.+ecl
<LI>34.,+54. ecl
<LI>34.%2B+54.%2Becl
</UL>
